### PR TITLE
 Added the ability to pass the test, also added statistics on the test

### PR DESCRIPTION
**Pull request by annalitvin**
Leader Bord in the process of change.

### CODE REVIEW

File: src/testset/views.py [hunk 1]
```diff
@@ -21,18 +29,118 @@ def get_context_data(self, *, object_list=None, **kwargs):
...
+                if qs_item[0].get('is_correct__count') == qs_item[1]:
+                    is_correct += 1
+
+            self.variants_count.clear()
+
+            current_test_result.finish()
+            current_test_result.save()
+            current_user = User.objects.get(pk=request.user.pk)
+            current_user.update_score()
+            current_user.save()
```
Comment (dbradul) [inline src/testset/views.py:117]: Consider subscribing on `test_result` save signal in User app and updating the `avr_score` attribute when test_result is completed (i.e. saved with `is_completed`=`True`)

Comment (dbradul) [reply src/testset/views.py:117]: Also, was idea of `avr_score` in user was its average score, not the last test result. 
But it's up to you if you how to implement it.

File: src/testset/templates/testset/test_end.html [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                      <tr>
+                        <th>{{ test.title }}</th>
+                        <th></th>
+                      </tr>
+                    </thead>
+                    <tbody>
+                      <tr>
+                        <td>Correct answers</td>
+                          {% with is_correct=is_correct  questions_count=questions_count %}
+                        <td> {{ is_correct }} of {{ questions_count }} ({{ is_correct|div:questions_count|mul:100 }})</td>
```
[the remark this answers is not in the document]
Comment (annalitvin) [reply src/testset/templates/testset/test_end.html:25]: fixed!

[the remark this answers is not in the document]
Comment (annalitvin) [reply src/testset/templates/testset/test_end.html:25]: fixed!

File: src/testset/views.py [hunk 3]
```diff
@@ -21,18 +29,118 @@ def get_context_data(self, *, object_list=None, **kwargs):
...
+        else:
+            qs = current_test_result.test_result_details.values('question').filter(is_correct=True).\
+                annotate(Count('is_correct'))
+
+            is_correct = 0
+            for qs_item in zip(qs, self.variants_count):
+                if qs_item[0].get('is_correct__count') == qs_item[1]:
+                    is_correct += 1
+
+            self.variants_count.clear()
```
Comment (dbradul) [inline src/testset/views.py:111]: As a part of my another comment, consider placing this into model. Actually, this is already done as a part of `update_score` method, you just need to create yet another representation of this data. Something like this, for example:
```
class TestResult(models.Model):
    ...
    def correct_answers_count(self):
        qs = self.details.values('question').annotate(
            num_answers=Count('question'),
            points=Sum('is_correct')
        )
        return sum(e['num_answers'] == int(e['points']) for e in qs)

    def score_info(self):
        num_questions = self.test.questions_count()
        num_answers = self.correct_answers_count()
        return f'{self.num_answers} of {num_questions} ({(self.num_answers/num_questions)*100:.2f}%)'
```

File: src/testset/views.py (hunk 1 above)
Comment (annalitvin) [reply src/testset/views.py:117]: > Also, was idea of `avr_score` in user was its average score, not the last test result.
> But it's up to you if you how to implement it.

fixed

File: src/testset/views.py (hunk 3 above)
Comment (annalitvin) [reply src/testset/views.py:111]: Fixed.

File: src/testset/models.py [hunk 4]
```diff
@@ -94,9 +92,13 @@ def update_score(self):
             for entry in qs
         )
 
+    def percent_correct_answer(self, is_correct, questions_count):
+        return is_correct / questions_count * 100
```
[the remark this answers is not in the document]
Comment (annalitvin) [reply src/testset/models.py:96]: Fixed.

**Pull request merged**